compiler/rustc_middle/src/ty/query.rs

```diff
@@ -227,8 +244,8 @@ macro_rules! define_callbacks {
                 pub fn $name<'tcx>(
                     _tcx: TyCtxt<'tcx>,
                     value: query_provided::$name<'tcx>,
-                ) -> query_values::$name<'tcx> {
-                    query_if_arena!([$($modifiers)*]
+                ) -> Erase<query_values::$name<'tcx>> {
+                    erase(query_if_arena!([$($modifiers)*]
```

Comment:
Could you add well-typed erase and restore functions for each query? Like `mod erase { fn $name ... }`?
That will be easier to reason about than manual erasure.

Reply:
Further: is there a way to use RPITs or TAITs to force some type safety, while keeping the small number of actual types for monomorphisation?

Reply:
Yeah I think using TAITs can work here.

Reply:
I ran into drop check issues with TAITs presumably since `rustc` doesn't know that `impl Copy` does not have a destructor.

Reply:
I fixed that with `ManuallyDrop`.